server/src/main/java/org/elasticsearch/cluster/coordination/LeaderChecker.java

```diff
@@ -0,0 +1,268 @@
+
+        @Override
+        public void close() {
+            if (isClosed.compareAndSet(false, true) == false) {
+                logger.debug("already closed");
+            } else {
+                logger.debug("closed");
+            }
+        }
+
+        void handleWakeUp() {
+            if (isClosed.get()) {
+                logger.debug("closed check scheduler woken up, doing nothing");
+                return;
+            }
+
+            logger.trace("checking {} with [{}] = {}", leader, LEADER_CHECK_TIMEOUT_SETTING.getKey(), leaderCheckTimeout);
+
+            // TODO lag detection:
+            // In the PoC, the leader sent its current version to the follower in the response to a LeaderCheck, so the follower
+            // could detect if it was lagging. We'd prefer this to be implemented on the leader, so the response is just
+            // TransportResponse.Empty here.
+            transportService.sendRequest(leader, LEADER_CHECK_ACTION_NAME, new LeaderCheckRequest(transportService.getLocalNode()),
+                TransportRequestOptions.builder().withTimeout(leaderCheckTimeout).withType(Type.PING).build(),
+
+                new TransportResponseHandler<TransportResponse.Empty>() {
+                    @Override
+                    public void handleResponse(Empty response) {
+                        if (isClosed.get()) {
+                            logger.debug("closed check scheduler received a response, doing nothing");
+                            return;
+                        }
+
+                        failureCountSinceLastSuccess.set(0);
+                        scheduleNextWakeUp(); // logs trace message indicating success
+                    }
+
+                    @Override
+                    public void handleException(TransportException exp) {
+                        if (isClosed.get()) {
+                            logger.debug("closed check scheduler received a response, doing nothing");
+                            return;
+                        }
+
+                        long failureCount = failureCountSinceLastSuccess.incrementAndGet();
+                        if (failureCount >= leaderCheckRetryCount) {
+                            logger.debug(new ParameterizedMessage("{} consecutive failures (limit [{}] is {}) so leader has failed",
+                                failureCount, LEADER_CHECK_RETRY_COUNT_SETTING.getKey(), leaderCheckRetryCount), exp);
+                            leaderFailed();
+                        } else {
+                            logger.debug(new ParameterizedMessage("{} consecutive failures (limit [{}] is {})",
+                                failureCount, LEADER_CHECK_RETRY_COUNT_SETTING.getKey(), leaderCheckRetryCount), exp);
+                            scheduleNextWakeUp();
+                        }
+                    }
+
+                    @Override
+                    public String executor() {
+                        return Names.GENERIC;
```

Comment:
should this be Names.SAME and spawn `onLeaderFailure.run();` to a new generic thread?

Reply:
I think 73bac5e759b22180ac4d57686a579bf18a6f03d4 was the fix for this.